model/task_blocked.go

```diff
@@ -0,0 +1,252 @@
+package model
+
+import (
+	"strings"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/tychoish/tarjan"
+)
+
+const (
+	taskBlocked  = "blocked"
+	taskPending  = "pending"
+	taskRunnable = "runnable"
+)
+
+// BlockedState returns "blocked," "pending" (unsatisfied dependencies,
+// but unblocked), or "" (runnable) to represent the state of the task
+// with respect to its dependencies
+func BlockedState(t *task.Task, tasksWithDeps []task.Task) (string, error) {
+	if t.DisplayOnly {
+		return blockedStateForDisplayTask(t, tasksWithDeps)
+	}
+	if t.IsPartOfSingleHostTaskGroup() {
+		return blockedStateForTaskGroups(t)
+	}
+	return blockedStatePrivate(t)
+}
+
+func blockedStatePrivate(t *task.Task) (string, error) {
+	if len(t.DependsOn) == 0 {
+		return taskRunnable, nil
+	}
+	dependencyIDs := []string{}
+	for _, d := range t.DependsOn {
+		dependencyIDs = append(dependencyIDs, d.TaskId)
+	}
+	dependentTasks, err := task.Find(task.ByIds(dependencyIDs).WithFields(task.DisplayNameKey, task.StatusKey,
+		task.ActivatedKey, task.BuildVariantKey, task.DetailsKey, task.DependsOnKey))
+	if err != nil {
+		return "", errors.Wrap(err, "error finding dependencies")
+	}
+	taskMap := map[string]*task.Task{}
+	for i := range dependentTasks {
+		taskMap[dependentTasks[i].Id] = &dependentTasks[i]
+	}
+	for _, dependency := range t.DependsOn {
+		depTask := taskMap[dependency.TaskId]
+		state, err := getStateByDependency(depTask, dependency)
+		if err != nil {
+			return "", errors.Wrap(err, "")
+		}
+		if state != taskRunnable {
+			return state, nil
+		}
+	}
+	return taskRunnable, nil
+}
+
+// getStateByDependency determines if the task is still running, is blocked, or has violated the given dependency
+func getStateByDependency(t *task.Task, dependency task.Dependency) (string, error) {
+	if t == nil {
+		grip.Error(message.Fields{
+			"message": "task does not exist",
+			"task_id": dependency.TaskId,
+		})
+		return taskRunnable, nil
+	}
+	state, err := blockedStatePrivate(t)
+	if err != nil {
+		return "", err
+	}
+	if state == taskBlocked {
+		return taskBlocked, nil
+	} else if t.Status == evergreen.TaskSucceeded || t.Status == evergreen.TaskFailed {
+		if t.Status != dependency.Status && dependency.Status != AllStatuses {
+			return taskBlocked, nil
+		}
+	} else {
+		return taskPending, nil
+	}
+	return taskRunnable, nil
+}
+
+func blockedStateForTaskGroups(t *task.Task) (string, error) {
```

Comment:
could we push this into an aggregation using similar code to the taskfinder?

Reply:
Sorry, I looked in taskFinder but I'm not sure what you mean. This logic replicates blockedStateForDisplayTask which is directly underneath.

Reply:
I was thinking about this: https://github.com/evergreen-ci/evergreen/blob/master/model/task/task.go#L1263

Reply:
This.. is pretty involved and probably out of scope haha. I'm sure it would be faster but this would probably be a change for all the functions in this file (i.e. BlockedStatePrivate might be the most useful to make an aggregation). Since people want this now I would put this off until later.

Reply:
cool, opening a ticket for it and marking it neweng might be cool?

Reply:
You want to make a new eng learn aggregations? :)